TOOLS/lua/autoload.lua

```diff
@@ -190,32 +188,43 @@ function find_and_add_entries()
     msg.trace("current file position in files: "..current)
 
     local append = {[-1] = {}, [1] = {}}
-    local filenames = get_playlist_filenames()
+    local filenames = get_playlist_filenames(pl)
     for direction = -1, 1, 2 do -- 2 iterations, with direction = -1 and +1
         for i = 1, MAXENTRIES do
-            local file = files[current + i * direction]
+            local pos = current + i * direction
+            local file = files[pos]
             if file == nil or file[1] == "." then
                 break
             end
 
             local filepath = dir .. file
             -- skip files already in playlist
-            if filenames[file] then break end
-
-            if direction == -1 then
-                if pl_current == 1 then -- never add additional entries in the middle
+            if not filenames[file] then
+                if direction == -1 then
                     msg.info("Prepending " .. file)
-                    table.insert(append[-1], 1, filepath)
+                    table.insert(append[-1], 1, {filepath, pos - 1})
```

Comment:
Quick question, why did you modify this to include the index as part of the table? I'm exactly clear on this. Is `add_files` ever expected to deal with indexes other than `pl_current + 1` and `pl_current`?

Reply:
This is because we support autoloading new files that weren't present in the initial run.
Some context: `add_files` is used for old-style playlist insertion that takes two commands, only on subsequent runs to insert files that didn't exist in the first run.  
The way this *used* to work (with `if filenames[file] then break end`) was that only new files that would sort past the current entry would get added, once playback reached the last playlist entry.  
This confusing behavior meant `pl_current + 1` was the only possible insert index after the initial run.  
Since we now support adding new files that will get sorted at any point in the playlist, the index at which they need to be inserted depends on the item.

Imagine playing file 2.mp4, which leads to an autoloaded playlist with a.mp4, c.mp4, d.mp4, then two new files b.mp4, e.mp4 get added, and should lead to a playlist with a.mp4, b.mp4, c.mp4, d.mp4, e.mp4. b and e were inserted at different points of the playlist. The old code only worked because of the 'known filenames' limitation (and arguably, a bug).

Hope this wasn't too confusing. you can easily compare the old vs new behavior by following the scenario described above, and observe that the old code would only add files past the current position, when we reached the end of the playlist, thus missing b.mp4.  
Another way to look at it is that the old code would ignore any files that sort before known files.

Reply:
Ah okay, that make sense. I was just wondering what the reasoning was.